### ISSUE TITLE
BatchNormalization use of Nd4j.createuninitialized() cuts performance by 30% on GPU

### ISSUE THREAD

**Issue by crockpotveggies** [deeplearning4j/deeplearning4j#3021]
Developed an Inception network of 13M parameters. A FaceNet variant. Adding batch norm layers to every convolution layer will cut throughput by 30%. Removing it creates a significant speedup.

Examining the profiler, the culprit appears to be coming from excessive calling of  Nd4j.createuninitialized() here:

https://github.com/deeplearning4j/deeplearning4j/blob/1f8af820c29cc5567a2c5eaa290f094c4d1492a7/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/normalization/BatchNormalization.java#L181

A high-level investigation points to either buffer or mem usage. Suggest @AlexDBlack and @raver119 have a look.

I suspect this same issue is applicable to many other layers, and many performance gains are to be had. I'm going to try the 2d vs. 4d "one weird trick" as per @huitseeker advice.

**Comment by raver119** [comment]
As explained in gitter, performance impact has nothing to do with createUninitialized.

It's either non-ews access (more likely), or strided access.

**Comment by crockpotveggies** [comment]
As per gitter conversation, initial culprits for non-ews access are:

```
org.deeplearning4j.nn.layers.normalization.BatchNormalization.backpropGradient:173
org.deeplearning4j.nn.layers.convolution.subsampling.SubsamplingLayer.backpropGradient:238
start here
org.nd4j.linalg.activations.impl.ActivationReLU.backprop:26
org.deeplearning4j.nn.layers.normalization.BatchNormalization.activate:231
```

**Comment by raver119** [comment]
https://gist.github.com/crockpotveggies/b3cb9d981cfc339f0af7492488925020

**Comment by raver119** [comment]
Any ideas where we at with this issue?
